mcf-browser/src/back-end/graph.js

```diff
@@ -0,0 +1,318 @@
+   */
+  async setExistsInKG() {
+    if (!this.dcid || this.existsInKG) {
+      return;
+    }
+    this.existsInKG = await getExistsInKG(this.dcid);
+  }
+
+  /**
+   * Creates Assertion objects from a list of property labels by calling the
+   * helper function getRemotePropertyValues from utils.js to find the values
+   * in Data Commons given the current node, a property label, and the direction
+   * of the label.
+   *
+   * @param {Array<string>} propLabels List of property labels associated with
+   *     the calling Node object in Data Commons.
+   * @param {boolean} isInverse True if the list of labels are incoming labels,
+   *     meaning the calling Node object is the target of the triple. False if
+   *     the calling Node is the source of the triple.
+   */
+  async createAssertionsFromLabels(propLabels, isInverse) {
+    for (const label of propLabels) {
+      await getRemotePropertyValues(this.dcid, label, isInverse)
+          .then((valueList) => {
+            if (!valueList) {
+              throw new Error('No property values for dcid: ' + this.dcid +
+                              ' label: ' + label);
+            }
+
+            valueList.forEach((valueObj) => {
+              const val = getValueFromValueObj(valueObj);
+              const source = isInverse ? val : this;
+              const target = isInverse ? this : val;
+              new Assertion(source, label, target, valueObj.provenanceId);
+            });
+          });
+    }
+  }
+
+  /**
+   * Stores remote triples as assertions and inverse Assertions of the calling
+   * Node object. Sets the alreadyFetched property to true if data is fetched.
+   */
+  async fetchRemoteData() {
+    if (this.alreadyFetched || !this.dcid) {
+      return;
+    }
+
+    await getRemotePropertyLabels(this.dcid).then(async (allLabels) => {
+      await this.createAssertionsFromLabels(allLabels.outLabels,
+                                            /* isInverse */ false);
+      await this.createAssertionsFromLabels(allLabels.inLabels,
+                                            /* isInverse */ true);
+    });
+    this.alreadyFetched = true;
+  }
+
+  /**
+   * Returns the linked list of assertions of the calling Node as an Array.
+   * @return {Array<Assertion>} The array of assertions.
+   */
+  getAssertions() {
+    const assertList = [];
+    let assert = this.assertions;
+
+    while (assert) {
+      assertList.push(assert);
+      assert = assert.nextAssertion;
+    }
+    return assertList;
+  }
+
+  /**
+   * Returns the linked list of inverse assertions of the calling Node as an
+   * Array.
+   * @return {Array<Assertion>} The array of Inverse assertions.
+   */
+  getInvAssertions() {
+    const invAssertList = [];
+
+    let invAssert = this.invAssertions;
+    while (invAssert) {
+      invAssertList.push(invAssert);
+      invAssert = invAssert.invNextAssertion;
+    }
+    return invAssertList;
```

Comment:
Based on what you have so far, I'm not seeing value in making Assertion a linked-list. I thought the same assertion object gets linked as fwd assertion in the src node, and inv assertion in the target node, but doesn't look like that's the case. Is there a reason I'm missing?

If not, feels like we'll simplify code by making assertions/inv-assertions just lists in Node.

Reply:
I think we were intending to use a linked list so that triples could render one at a time as they were remotely fetched, instead of having to wait for all of the triples to be fetched before rendering all of them at once. I ended up not using the approach when implementing the front-end (maybe an optimization that could be made in the future?), so I can switch to using arrays instead. If we end up wanting this rendering, then I can switch back to link lists at that point.